internal/testutil/all_validators_test.go

```diff
@@ -0,0 +1,47 @@
+package testutil
+
+import (
+	"testing"
+
+	"github.com/mt-sre/addon-metadata-operator/pkg/utils"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type Validator interface {
+	Name() string
+	Run(utils.MetaBundle) (bool, error)
+	SucceedingCandidates() []utils.MetaBundle
+	FailingCandidates() []utils.MetaBundle
+}
+
+// register the validators to test here by appending them to the following slice
+var (
+	validatorsToTest []Validator = []Validator{
+		ValidatorAddonLabelTestBundle{},
+		ValidatorDefaultChannelTestBundle{},
+	}
+)
+
+func Test_AllValidators(t *testing.T) {
+	for _, validator := range validatorsToTest {
+		validator := validator
+		t.Run(validator.Name(), func(t *testing.T) {
+			t.Parallel()
+			// testing the succeeding candidates
+			succeedingMetaBundles := validator.SucceedingCandidates()
```

Comment:
We could thread this even more aggressively. Since goroutines are pretty small constructs, it would probably provide a small performance increase (although we should measure it). This is prematured optimization but seems quite straightforward. Since we can use the testing framework to report errors it's also ultra simple and we dont need to use an error channel or anything like that.

```go
succeedingMetaBundles := validator.SucceedingCandidates()
failingMetaBundles := validator.FailingCandidates()

var wg sync.WaitGroup
wg.Add(len(succeedingMetaBundles) + len(failingMetaBundles))

for _, mb := range succeedingMetaBundles {
  go func(v Validator, mb utils.MetaBundle) {
    success, err := v.Run(mb)
    require.NoError(t, err)
    assert.True(t, success)
    wg.Done()
  }(validator, mb)
}

for _, mb := range failingMetaBundles {
  go func(v Validator, mb utils.MetaBundle) {
    success, err := v.Run(mb)
    require.NoError(t, err)
    assert.False(t, success)
    wg.Done()
  }(validator, mb)
}
wg.Wait()
```

Reply:
My only question would be, if one of the goroutine fails, are the other goroutines stopped, garbage collected? We would want an early exit.

Reply:
Yep thought of this too but realised that the computation of these tests (at least until now) isn't I/O bound/blocking. They are simply CPU-bound non-blocking tasks, so I don't think that "cumulatively" the use of goroutines endup providing any performance benefits. 
Also, if in the future, our piece of code ends up generating, say, a 500 goroutines and if all of them are non-blocking only, then, they would actually end up imposing an increase in latency as well considering the context switches between 500 goroutines. Although the latency increase with N goroutines wouldn't be as substantial as when faced with N number of OS threads considering that the goroutines end-up multiplexing to one OS thread.

Also, the `t.Parallel()` is achieving a baseline parallelism/concurrency at a per-validator-test basis. The concurrency at a per-candidate basis seems a bit overkill to me and again, prone to spinning up a heck lot of goroutines and potential latency jumps.

That's what I think, would love to know your thoughts as well and would love to be corrected if I am wrong anywhere :)

Reply:
In case, we decide to proceed with implementing goroutines:

> My only question would be, if one of the goroutine fails, are the other goroutines stopped, garbage collected? We would want an early exit.

Regarding this, we can have an unbuffered channel called `yeet` (whatever). And the sender to that channel would be the piece of code executing the validator tests. Whenever a validator test fails, a value will be sent to this `yeet` channel.
The receiving end of this channel would be simply a separate goroutine which will be running an infinite loop and constantly checking if it received any value from the `yeet` channel. And whenever it will receive a value from the `yeet` channel, it will just `os.Exit(1)` our program.

So, in simple words, the flow will be:
- a common unbuffered channel of communication.
- Our validator tests run and the first time they encounter a failure, the send a value to that channel.
- In real-time, the receiver goroutine captures a correspondingly from that channel and takes it as an indication to run `os.Exit(1)`

PS: the only thing I want to confirm about is the garbage collection of other goroutines which will be under execution, when os.Exit(1) will happen

Reply:
>the only thing I want to confirm about is the garbage collection of other goroutines which will be under execution, when os.Exit(1) will happen
There is no garbage collection to be done after you call `os.Exit()` the process itself exits.

Reply:
@sugarraysam  I also agree with @yashvardhan-kukreja, I think scaling the number of go-routines linearly with the amount of validator tests to run wouldn't give us any extra performance benefits. I think the best approach here is to stick with `t.Parallel()`.

Reply:
Perfect. Thanks for your answer. I agree it is overkill for now. Let's revisit if there is ever a performance increase need.